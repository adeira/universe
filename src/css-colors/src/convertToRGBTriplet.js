// @flow strict

import { sprintf } from '@adeira/js';

import cssColorNames from './cssColorNames';
import hex3ToHex6 from './hex3ToHex6';
import isColor from './isColor';
import { RGBA_PATTERN_MATCH } from './utils/isRGBA';

const HEX_REGEXP_SHORT = /^#(?<R>[a-f0-9])(?<G>[a-f0-9])(?<B>[a-f0-9])$/i;
const HEX_REGEXP_LONG = /^#(?<R>[a-f0-9]{2})(?<G>[a-f0-9]{2})(?<B>[a-f0-9]{2})$/i;

/**
 * Tries to convert the input value into RGB triplet which is commonly used throughout this library
 * for further calculations (brightness, accessibility, â€¦). The input value can be anything that is
 * valid in CSS/HTML.
 *
 * See: https://developer.mozilla.org/en-US/docs/Web/CSS/color_value
 *
 * It throws an error when the conversion is not implemented yet.
 *
 * TODO: How to deal with alpha channel (not supported in the triplet). Should we introduce custom
 *  object to keep all the extracted values?
 */
export default function convertToRGBTriplet(value: string): [number, number, number] {
  if (isColor(value)) {
    const rgbaColorMatch = value.match(RGBA_PATTERN_MATCH);
    if (cssColorNames.has(value)) {
      const hexColor = cssColorNames.get(value);
      // $FlowIssue[incompatible-call]: `hexColor` should not be undefined after calling `has` above
      return convertHexToRGBTriplet(hexColor);
    } else if (HEX_REGEXP_LONG.test(value)) {
      return convertHexToRGBTriplet(value);
    } else if (HEX_REGEXP_SHORT.test(value)) {
      return convertHexToRGBTriplet(hex3ToHex6(value));
    } else if (rgbaColorMatch) {
      return convertRGBToTriplet(
        rgbaColorMatch.groups?.R,
        rgbaColorMatch.groups?.G,
        rgbaColorMatch.groups?.B,
      );
    }
    throw new Error(sprintf('Conversion of color "%s" is currently not supported.', value));
  } else {
    throw new Error(sprintf('Value "%s" is not a color!', value));
  }
}

function convertHexToRGBTriplet(hexColor: string): [number, number, number] {
  // TODO: we currently ignore alpha channels of the #RGB colors (#RGBA)
  const hexColorMatch = hexColor.match(HEX_REGEXP_LONG);
  return [
    parseInt(hexColorMatch?.groups?.R, 16),
    parseInt(hexColorMatch?.groups?.G, 16),
    parseInt(hexColorMatch?.groups?.B, 16),
  ];
}

function convertRGBToTriplet(r, g, b): [number, number, number] {
  return [
    Number(r?.endsWith('%') ? convertPercentageToNumber(r) : r),
    Number(g?.endsWith('%') ? convertPercentageToNumber(g) : g),
    Number(b?.endsWith('%') ? convertPercentageToNumber(b) : b),
  ];
}

function convertPercentageToNumber(percentage) {
  const match = percentage?.match(/^(?<number>\d+)%$/)?.groups;
  return Math.round((255 / 100) * Number(match?.number));
}
