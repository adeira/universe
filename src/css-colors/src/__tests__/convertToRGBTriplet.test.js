// @flow strict

import convertToRGBTriplet from '../convertToRGBTriplet';
import { validColors } from './fixtures/validColors';

test.each([
  /* $FlowFixMe[incompatible-call] This comment suppresses an error when
   * upgrading Flow to version 0.199.1. To see the error delete this comment
   * and run Flow. */
  ...validColors.keywords.map(([color, colorTriplet]) => [color, colorTriplet]),
  ...validColors.hexadecimal.map(([color, colorTriplet]) => [color, colorTriplet]),
  ...validColors.functional.map(([color, colorTriplet]) => [color, colorTriplet]),
])('converts %s => %s', (color, colorTriplet) => {
  if (colorTriplet === null) {
    expect(() => convertToRGBTriplet(color)).toThrowErrorMatchingSnapshot();
  } else {
    expect(convertToRGBTriplet(color)).toEqual(colorTriplet);
  }
});
