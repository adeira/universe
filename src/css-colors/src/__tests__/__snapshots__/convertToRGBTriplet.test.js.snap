// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts #F09F => null 1`] = `"Conversion of color "#F09F" is currently not supported."`;

exports[`converts #FF0099FF => null 1`] = `"Conversion of color "#FF0099FF" is currently not supported."`;

exports[`converts #f09f => null 1`] = `"Conversion of color "#f09f" is currently not supported."`;

exports[`converts #ff0099ff => null 1`] = `"Conversion of color "#ff0099ff" is currently not supported."`;

exports[`converts currentcolor => null 1`] = `"Conversion of color "currentcolor" is currently not supported."`;

exports[`converts hsl(.75turn, 60%, 70%) => null 1`] = `"Conversion of color "hsl(.75turn, 60%, 70%)" is currently not supported."`;

exports[`converts hsl(4.71239rad, 60%, 70%) => null 1`] = `"Conversion of color "hsl(4.71239rad, 60%, 70%)" is currently not supported."`;

exports[`converts hsl(270 60% 50% / .15) => null 1`] = `"Conversion of color "hsl(270 60% 50% / .15)" is currently not supported."`;

exports[`converts hsl(270 60% 50% / 15%) => null 1`] = `"Conversion of color "hsl(270 60% 50% / 15%)" is currently not supported."`;

exports[`converts hsl(270 60% 50%/.15) => null 1`] = `"Conversion of color "hsl(270 60% 50%/.15)" is currently not supported."`;

exports[`converts hsl(270 60% 70%) => null 1`] = `"Conversion of color "hsl(270 60% 70%)" is currently not supported."`;

exports[`converts hsl(270, 60%, 50%, .15) => null 1`] = `"Conversion of color "hsl(270, 60%, 50%, .15)" is currently not supported."`;

exports[`converts hsl(270, 60%, 50%, 15%) => null 1`] = `"Conversion of color "hsl(270, 60%, 50%, 15%)" is currently not supported."`;

exports[`converts hsl(270, 60%, 70%) => null 1`] = `"Conversion of color "hsl(270, 60%, 70%)" is currently not supported."`;

exports[`converts hsl(270deg, 60%, 70%) => null 1`] = `"Conversion of color "hsl(270deg, 60%, 70%)" is currently not supported."`;

exports[`converts hsla(.75turn, 60%, 70%) => null 1`] = `"Conversion of color "hsla(.75turn, 60%, 70%)" is currently not supported."`;

exports[`converts hsla(4.71239rad, 60%, 70%) => null 1`] = `"Conversion of color "hsla(4.71239rad, 60%, 70%)" is currently not supported."`;

exports[`converts hsla(270 60% 50% / .15) => null 1`] = `"Conversion of color "hsla(270 60% 50% / .15)" is currently not supported."`;

exports[`converts hsla(270 60% 50% / 15%) => null 1`] = `"Conversion of color "hsla(270 60% 50% / 15%)" is currently not supported."`;

exports[`converts hsla(270 60% 70%) => null 1`] = `"Conversion of color "hsla(270 60% 70%)" is currently not supported."`;

exports[`converts hsla(270, 60%, 50%, .15) => null 1`] = `"Conversion of color "hsla(270, 60%, 50%, .15)" is currently not supported."`;

exports[`converts hsla(270, 60%, 50%, 15%) => null 1`] = `"Conversion of color "hsla(270, 60%, 50%, 15%)" is currently not supported."`;

exports[`converts hsla(270, 60%, 70%) => null 1`] = `"Conversion of color "hsla(270, 60%, 70%)" is currently not supported."`;

exports[`converts hsla(270deg, 60%, 70%) => null 1`] = `"Conversion of color "hsla(270deg, 60%, 70%)" is currently not supported."`;

exports[`converts transparent => null 1`] = `"Conversion of color "transparent" is currently not supported."`;
